packages/core/src/render3/hooks.ts

```diff
@@ -158,25 +189,51 @@ export function executeHooks(
  *
  * @param currentView The current view
  * @param arr The array in which the hooks are found
+ * @param initPhaseState the current state of the init phase
+ * @param currentNodeIndex 3 cases depending the the value:
+ * - undefined: all hooks from the array should be executed (post-order case)
+ * - null: execute hooks only from the saved index util the end of the array (pre-order case, when
+ * flushing the remaining hooks)
+ * - number: execute hooks only from the saved index util that node index exclusive (pre-order case,
+ * when executing flushHooksUpTo(number))
  */
-export function callHooks(currentView: LView, arr: HookData, initPhase?: number): void {
-  let initHooksCount = 0;
-  for (let i = 0; i < arr.length; i += 2) {
-    const isInitHook = arr[i] < 0;
-    const directiveIndex = isInitHook ? -arr[i] : arr[i] as number;
-    const directive = currentView[directiveIndex];
+function callHooks(
+    currentView: LView, arr: HookData, initPhase: InitPhaseState,
+    currentNodeIndex?: number | null): void {
+  const startIndex = currentNodeIndex !== undefined ?
+      (currentView[FLAGS_MORE] & LViewFlagsMore.IndexOfTheNextPreOrderHookMaskMask) :
+      0;
+  const nodeIndexLimit = currentNodeIndex != null ? currentNodeIndex : -1;
+  let lastNodeIndexFound = 0;
+  for (let i = startIndex; i < arr.length; i += 2) {
     const hook = arr[i + 1] as() => void;
-    if (isInitHook) {
-      initHooksCount++;
-      const indexWithintInitPhase = currentView[FLAGS] >> LViewFlags.IndexWithinInitPhaseShift;
-      // The init phase state must be always checked here as it may have been recursively updated
-      if (indexWithintInitPhase < initHooksCount &&
-          (currentView[FLAGS] & LViewFlags.InitPhaseStateMask) === initPhase) {
-        currentView[FLAGS] += LViewFlags.IndexWithinInitPhaseIncrementer;
-        hook.call(directive);
+    if (hook === NOOP) {
```

Comment:
What do you think about storing the node index alone (un-encoded and without `NOOP`)?  We can tell that an index is an element index if it's followed by another number, rather than a function. That way, we wouldn't have to allocate storage space for `NOOP`?

```suggestion
    if (typeof hook === 'number') {
```